packages/fxa-settings/src/components/PageTwoStepAuthentication/index.tsx

```diff
@@ -56,73 +75,104 @@ export const PageTwoStepAuthentication = (_: RouteComponentProps) => {
+                alt={`Use the code ${secret} to set up two-step authentication in supported applications.`}
+              />
+            )}
+          </div>
+
+          <p className="mt-4">
+            Now enter the security code from the authentication app.
+          </p>
+
+          <div className="mt-4 mb-6" data-testid="recovery-key-input">
+            <InputText
+              name="totp"
+              label="Enter security code"
+              prefixDataTestId="totp"
+              maxLength={6}
+              autoFocus
+              onChange={() => {
+                setInvalidCodeError('');
+                trigger('totp');
+              }}
+              inputRef={register({
+                validate: isValidTotpFormat,
+              })}
+              {...{ errorText: invalidCodeError }}
             />
-          )}
-        </div>
-
-        <p className="mt-4">
-          Now enter the security code from the authentication app.
-        </p>
-
-        <div className="mt-4 mb-6" data-testid="recovery-key-input">
-          <InputText
-            name="totp"
-            label="Enter security code"
-            prefixDataTestId="totp"
-            maxLength={6}
-            onChange={() => trigger('totp')}
-            inputRef={register({
-              validate: isValidTotpFormat,
-            })}
-          />
-        </div>
-
-        <div className="flex justify-center mb-4 mx-auto max-w-64">
-          <button
-            type="button"
-            className="cta-neutral mx-2 flex-1"
-            onClick={() => window.history.back()}
-          >
-            Cancel
-          </button>
-          <button
-            type="submit"
-            data-testid="submit-totp"
-            className="cta-primary mx-2 flex-1"
-            disabled={!formState.isDirty || !formState.isValid}
-          >
-            Continue
-          </button>
-        </div>
-      </form>
+          </div>
+
+          <div className="flex justify-center mb-4 mx-auto max-w-64">
+            <button
+              type="button"
+              className="cta-neutral mx-2 flex-1"
+              onClick={goBack}
+            >
+              Cancel
+            </button>
+            <button
+              type="submit"
+              data-testid="submit-totp"
+              className="cta-primary mx-2 flex-1"
+              disabled={!formState.isDirty || !formState.isValid}
+            >
+              Continue
+            </button>
+          </div>
+        </form>
+      )}
+      {totpVerified && (
+        <form>
+          <div className="my-2" data-testid="2fa-recovery-codes">
+            Save these one-time use codes in a safe place for when you don’t
+            have your mobile device.
+            <div className="mt-6 flex flex-col items-center h-40 justify-between">
+              <DataBlock value={recoveryCodes}></DataBlock>
+              <GetDataTrio value={recoveryCodes}></GetDataTrio>
```

Comment:
I think we need to adjust the string formatting of `GetDataTrio` actions for the string array case. _Download_ for example gives me one long string

Reply:
We've found that https://github.com/mozilla/fxa/issues/4975 should handle that issue.